robot-server/robot_server/runs/run_store.py

```diff
@@ -32,12 +35,64 @@ def __init__(self, run_id: str) -> None:
 class RunStore:
     """Methods for storing and retrieving run resources."""
 
-    def __init__(self) -> None:
+    def __init__(self, sql_engine: sqlalchemy.engine.Engine) -> None:
         """Initialize a RunStore and its in-memory storage."""
-        self._runs_by_id: Dict[str, RunResource] = {}
+        self._sql_engine = sql_engine
 
-    def upsert(self, run: RunResource) -> RunResource:
-        """Insert or update a run resource in the store.
+    def update_actions(self, run_id: str, actions: List[RunAction]) -> None:
+        """Insert or update a run actions resource in the db.
+
+        Arguments:
+            run_id: current run id to get
+            actions: a list of actions to store in the db
+        """
+        try:
+            with self._sql_engine.begin() as transaction:
+                transaction.execute(
+                    sqlalchemy.delete(action_runs_table).where(
+                        action_runs_table.c.run_id == run_id
+                    )
+                )
+        except sqlalchemy.exc.NoResultFound as e:
+            raise e
+
+        self.insert_actions(run_id, actions)
+
+    def insert_actions(self, run_id: str, actions: List[RunAction]) -> None:
+        """Insert or update a run actions resource in the db.
+
+        Arguments:
+            run_id: current run id to get
+            actions: a list of actions to store in the db
+        """
+        insert_actions = []
+        for action in actions:
+            insert_actions.append(_convert_action_to_sql_values(action, run_id))
```

Comment:
Stylistically, we generally prefer list comprehensions over procedurally building up a list element-by-element.

```suggestion
        actions_to_insert = [
            _convert_action_to_sql_values(action, run_id) for action in actions
        ]
```